tox.ini

```diff
@@ -42,13 +42,8 @@ commands=
     native-state-byzantium: py.test {posargs:tests/json-fixtures/test_state.py -k Byzantium}
     native-state-constantinople: py.test {posargs:tests/json-fixtures/test_state.py -k Constantinople}
     native-state-metropolis: py.test {posargs:tests/json-fixtures/test_state.py -k Metropolis}
-extras =
-    coincurve
-    database: leveldb
-    trinity: trinity, p2p
-    rpc: trinity, p2p
-    p2p: p2p
-deps = -r{toxinidir}/requirements-dev.txt
+
+deps = .[dev]
```

Comment:
Yous can just change this to `extras=dev` which is the slightly more semantically correct way to do this.

Reply:
:+1:

Reply:
What was wrong with the extras as the were? They had a clean isolation of which dependencies to install for which tests.

Reply:
If we do it right, we shouldn't need to install the "developer" extra inside the test suite at all: https://github.com/ethereum/eth-account/blob/master/tox.ini#L32-L35

Reply:
One quirk that I never found an answer for: I'd *like* to specify `eth-hash` as a primary dependency in a library like `py-evm` and then specify an extra, such that `py-evm[test]` depends on `eth-hash[pycryptodome]`. The first few ways I've tried that didn't work as I expected, so I was stuck doing it as an explicit dependency outside the test extra, like:

https://github.com/ethereum/eth-utils/blob/03eabede3d55e2d00686d2b14bda0cb2148c63af/tox.ini#L14-L16

This is a tradeoff that seems like the better option, but does have a downside:
- Explicit `eth-hash[pycryptodome]` in `tox.ini`
  - Bad: We have to specify the backend of choice twice: once in the tox file and once in setup.py. Changing the preferred backend means changing it twice (or more realistically: changing it once by accident, and hunting down a confusing bug from two slightly different environments)
  - Good: We get to specify which major versions of pycryptodome that eth-hash supports **inside** eth-hash, so any changes in supported versions auto-propagate to all libraries that use eth-hash
- Specify `pycryptodome` version explicitly in `py-evm[test]`
  - Bad: if we need to change the supported versions of pycryptodome (say, they make a non-semver-compliant release, and we have to explicitly exclude some minor version), then all libraries that depend on eth-hash will be broken until we go tweak the lib.
  - Good: the backend is only specified once in `py-evm` so we eliminate the class of bugs from the first approach

Since a breakage in the first one only affects this one library (py-evm), and the second one affects *all* libraries depending on eth-hash, the first one seems the clear winner. (unless we have a strong reason to believe that one kind of breakage is orders of magnitude less likely than the other)

Reply:
>What was wrong with the extras as the were? They had a clean isolation of which dependencies to install for which tests.

I haven't had a good mental model of the dependencies before @pipermerriam explained it further. I'm still probably lacking some important details, but here comes my second take nevertheless.

Let's start with this: I changed the `install_requires` to 

```
install_requires = deps['evm'] + deps['p2p'] + deps['trinity']
```

Because I believe anyone who runs `pip install py-evm` does expect this to install `evm`, `p2p` and `trinity` fully to the point that no further manual installs are needed. At least for now that `p2p` and `trinity` are part of this package.

That's why extras such as `p2p` or `trinity` are gone because they are simply installed as part of the regular dependencies.

However, based on the feedback that I received I separated out a new section `evm-extras`

```
    'evm-extras': [
        "coincurve>=7.0.0,<8.0.0",
        "plyvel==1.0.4",
        "eth-hash[pycryptodome]",
    ]
``` 
Previously these were separate sections (`coincurve` and `leveldb`) and  `"eth-hash[pycryptodome]"` was living under `requirements-dev.txt` (which is now gone). It made sense to me to group these under one section but maybe there are good reasons to have them separately? I think it's more convenient to have them grouped and yet people can still install separately whatever they choose.

>In the ideal setup, we shouldn't need to install the "developer" extra inside the test suite at all

Yes, I agree and in fact, the tests now only need `evm-extras` and `test` as extras whereas the linting jobs just need `lint` as extras.
   
>One quirk that I never found an answer for: I'd like to specify eth-hash as a primary dependency in a library like py-evm and then specify an extra, such that py-evm[test] depends on eth-hash[pycryptodome]

Ok, that sounds odd indeed. So, if I understand you correctly you are saying that if you specify `eth-hash` as `install_requires` you can not overwrite it in an extra with `eth-hash[pycryptodome]`?

@pipermerriam 

>Yous can just change this to `extras=dev` which is the slightly more semantically correct way to do this.

This didn't seem to work for me. The `extras =...` option seems to only work in combination with `deps=...`. As soon as you don't have `deps` anymore the `extras` didn't seem to have an effect. It worked when I did `deps = .` in combination with `extras = ...` but that seems more verbose than just going for `deps = .[x, y, z]` straight away.

Alright, hope this update makes more sense :grimacing:

Reply:
@pipermerriam now that I refactored the CI runs to not depend on the `dev` dependencies anymore the trinity tests are missing `ipython`. Is trinity supposed to have a hard dependency against ipython by default (e.g. putting it into the `trinity` section directly) or should that go into something like `trinity-extras` (somewhat reasonable I believe since one can also run trinity with `--vanilla-shell`)?

Reply:
> I believe anyone who runs pip install py-evm does expect this to install evm, p2p and trinity fully to the point that no further manual installs are needed. At least for now that p2p and trinity are part of this package.

Fair enough. This mostly just makes me eager to split trinity out into its own repo for real (but we're not there right now).

> So, if I understand you correctly you are saying that if you specify eth-hash as install_requires you can not overwrite it in an extra with eth-hash[pycryptodome]?

That was my experience, but I don't know for sure. It was one of those things that didn't *just work* and I didn't have time to investigate if it was travis or tox or pip.

> As soon as you don't have deps anymore the extras didn't seem to have an effect.

That's super-weird and feels broken. Heck, that might have even been the cause of the previous problem. If that's trivially reproducible, that sounds like a fairly straightforward thing to file a bug against tox for.

Reply:
>That's super-weird and feels broken

Yep, in fact it really felt broken because it sometimes seemed to work for one run and then failed for another whereas `.deps[x,y,z]` worked reliably every single time.

>Heck, that might have even been the cause of the previous problem

Yes, that might be it.

>If that's trivially reproducible, that sounds like a fairly straightforward thing to file a bug against tox for

I can investigate that further *another time*  :grimacing: